### PR TITLE
Implemented search algorithms and their test

### LINKED ISSUE #23: Implement searching algorithms
Add these algorithms
- [x] 1. Linear Search
- [x] 2. Binary Search
- [x] 3. Jump Search
- [x] 4. Interpolation Search
- [ ] 5. Exponential Search
- [x] 6. Fibonacci Search
- [ ] 7. Recursive program to linearly search an element in a given array
- [ ] 8. Recursive function to do a substring search

Also write test cases for the same. Tests and code can be divided in two commits though.

### PR DESCRIPTION
**Pull request by nisheksharma**
### Description
Implemented different search algorithms and their respective tests as follows

1. Linear Search
2. Binary Search
3. Jump Search
4. Interpolation Search
5. Fibonacci Search

 

### Related Issue

** Please don't create a Pull request without creating an issue first **

Fixes #23

### Type of Change:

- Code
- Quality Assurance
- Documentation

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)
- This change requires a documentation update (software upgrade on readme file)

### How Has This Been Tested?
Unittest

### Checklist:

- [x] I read the **CONTRIBUTING** document before making this pull request
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] My change requires change to documentation
- [x] I have updated documentation accordingly
- [x] I have updated the requirements file
- [x] My changes generate no new warnings 
- [x] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,2 @@
-# dsalgo
\ No newline at end of file
+# dsalgo
```
Comment (codesankalp) [inline README.md:1]: no need to update the readme.
check our new readme.

File: dsalgo/search.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
+import math
+class Search:
+
+    # Iterative Binary Search Function
```
Comment (codesankalp) [inline dsalgo/search.py:4]: add docstrings.

File: dsalgo/search.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    
+            # If x is smaller, ignore right half 
+            else: 
+                return mid 
+    
+        # If we reach here, then the element was not present 
+        return -1
+    # Searching an element in a list/array in python 
+    # can be simply done using \'in\' operator 
+    # Example:
```
Comment (codesankalp) [inline dsalgo/search.py:32]: no need to write comments to explain because no one is going to read this.

File: dsalgo/search.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    # else return -1 
+    
+    def linear_search(arr, x): 
+    
+        for i in range(len(arr)): 
+    
+            if arr[i] == x: 
+                return i 
+    
+        return -1
```
Comment (codesankalp) [inline dsalgo/search.py:49]: add lines between them
use flake 8

File: dsalgo/search.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            c=i
+            for j in arr[i:]:
+                if j==search:
+                    return c
+                c+=1
+        else:
+            arr_ls = arr[chunk-interval:chunk]
+            ind = [i for i,d in enumerate(arr_ls) if d==search]
+            return chunk-interval+ind[0]
+    # If x is present in arr[0..n-1], then returns
```
Comment (codesankalp) [inline dsalgo/search.py:70]: again wrong format.

File: dsalgo/search.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                    return c
+                c+=1
+        else:
+            arr_ls = arr[chunk-interval:chunk]
+            ind = [i for i,d in enumerate(arr_ls) if d==search]
+            return chunk-interval+ind[0]
+    # If x is present in arr[0..n-1], then returns 
+# index of it, else returns -1 
+    def interpolation_search(arr, length, x): 
+        # Find indexs of two corners
```
Comment (codesankalp) [inline dsalgo/search.py:73]: add docstring

File: dsalgo/search.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    
+            # If x is larger, x is in upper part 
+            if arr[pos] < x: 
+                lo = pos + 1; 
+    
+            # If x is smaller, x is in lower part 
+            else: 
+                hi = pos - 1; 
+        return -1
+if __name__ == "__main__":
```
Comment (codesankalp) [inline dsalgo/search.py:102]: no need for this,

File: dsalgo/search.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
+import math
+class Search:
+
+    # Iterative Binary Search Function 
+    # It returns index of x in given array arr if present, 
+    # else returns -1 
+    def binary_search(arr, x): 
+        low = 0
```
Comment (codesankalp) [inline dsalgo/search.py:8]: where is __init__

**Review by codesankalp — CHANGES_REQUESTED** [review]
You have to link your issue with pull request.
Try to do the mentioned changes.

**Comment by codesankalp** [comment]
also this branch have conflicts so rebase your branch with master,
try to squash commit into one

**Comment by codesankalp** [comment]
@nisheksharma  after 2 days if you don't change anything then it will be marked as invalid.

**Comment by nisheksharma** [comment]
@codesankalp review

**Review by codesankalp — APPROVED** [review]
Tested all the code and works properly.
Code styling is also good.
It is ready to merge.

**Comment by codesankalp** [comment]
@Aryamanz29 
@17sushmita 
please review this pr.

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
@nisheksharma rename your file to test_search

**Comment by devkapilbansal** [comment]
Also, please follow commit message guidelines

**Comment by devkapilbansal** [comment]
Pinging @17sushmita for testing code locally before merge

**Comment by devkapilbansal** [comment]
Thanks @nisheksharma just one last request from my side. Change type to feature instead of file, as file is not a commit type defined in our guidelines

**Review by codesankalp — APPROVED** [review]

**Comment by nisheksharma** [comment]
> @nisheksharma rename your file to test_search

Done

**Comment by nisheksharma** [comment]
> Thanks @nisheksharma just one last request from my side. Change type to feature instead of file, as file is not a commit type defined in our guidelines

done

**Review by devkapilbansal — APPROVED** [review]
@nishesksharma thanks Made some changes. Hope you don't mind

**Comment by devkapilbansal** [comment]
LGTM 👍 Thanks @nisheksharma

**Pull request merged**